File: spine/orreth_spine/envelope.py
# PROVENANCE: Claude Fable 5 (claude-fable-5) — rearch P0 sp1, the rig breathes · 2026-09-16
"""The shared transport envelope (orreth.transport/1), v0.

One broker-neutral metadata contract for every rail (canon 0002). An
envelope is encoded as canonical bytes — sorted keys, compact separators,
ASCII — so the same message is the same bytes on every rail and in every
language. Payloads are pointer-shaped by default: a ref and a hash, never
bodies or prompts. The authority chain says who created and who requested
(canon 0004: attribution is the point) — the origin human first, then
each delegating identity in order.
"""
from __future__ import annotations

import hashlib
import json
import os
import secrets
from datetime import datetime, timezone


def scope() -> str:
    """The world this process belongs to (SPINE_SCOPE, default u:dev).
    Facts wear it, and a dispatcher only dispatches its OWN world's
    facts — two rigs on one broker stay two worlds (found live: the
    running Bridge's dispatcher bridged a test's facts onto its own
    benches and its librarian answered them)."""
    return os.environ.get("SPINE_SCOPE", "u:dev")

SPECVERSION = "orreth.transport/1"
KINDS = ("command", "event")
REQUIRED = ("specversion", "message_id", "message_kind", "type",
            "universe_id", "scope_path", "occurred_at", "payload")


def canonical(obj) -> bytes:
    """The one true byte form: sorted keys, compact, ASCII."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=True).encode("ascii")


def content_hash(obj) -> str:
    """sha256 over the canonical bytes, prefixed so the algorithm is named."""
    return "sha256:" + hashlib.sha256(canonical(obj)).hexdigest()


def now_iso() -> str:
    """UTC, millisecond precision, Z-suffixed — one clock spelling everywhere."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def make_envelope(*, kind: str, type: str, universe_id: str, scope_path: str,
                  payload: dict, correlation_id: str | None = None,
                  authority_chain: list[str] | None = None,
                  aggregate: dict | None = None,
                  marker: dict | None = None) -> dict:
    """Mint one envelope. The message id is globally unique and immutable
    (commands wear cmd_, events wear msg_); everything else is plain data.
    `aggregate` ({type, id, sequence}) names the one thing whose order
    matters — sequences are monotonic PER AGGREGATE, never globally
    (canon 0002: there is no useful global order)."""
    if kind not in KINDS:
        raise ValueError(f"message_kind must be one of {KINDS}, not {kind!r}")
    env = {
        "specversion": SPECVERSION,
        "message_id": ("cmd_" if kind == "command" else "msg_")
                      + secrets.token_hex(12),
        "message_kind": kind,
        "type": type,
        "universe_id": universe_id,
        "scope_path": scope_path,
        "occurred_at": now_iso(),
        "payload": payload,
    }
    if correlation_id:
        env["correlation_id"] = correlation_id
    if authority_chain:
        env["authority_chain"] = list(authority_chain)
    if aggregate:
        env["aggregate"] = dict(aggregate)
    if marker:                      # the typed origin this fact serves (0006):
        env["marker"] = dict(marker)   # {kind, id, parent, by} — WHY, never who
    return env


def encode(env: dict) -> bytes:
    """Canonical bytes of a complete envelope; an incomplete one is refused
    with its missing fields named — never guessed at."""
    missing = [k for k in REQUIRED if k not in env]
    if missing:
        raise ValueError("envelope missing required fields: " + ", ".join(missing))
    if env["specversion"] != SPECVERSION:
        raise ValueError(f"unknown specversion {env['specversion']!r}")
    return canonical(env)


def decode(raw: bytes) -> dict:
    """Read an envelope back. Unknown additive fields are preserved (a newer
    publisher never breaks an older reader); missing required ones refuse."""
    env = json.loads(raw.decode("ascii"))
    missing = [k for k in REQUIRED if k not in env]
    if missing:
        raise ValueError("envelope missing required fields: " + ", ".join(missing))
    return env

File: backend/plane/crates/orreth-spine/src/envelope.rs
// PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp1, the bytes · 2026-09-22
//! The shared transport envelope (`orreth.transport/1`), v0 — mirrors
//! `orreth_spine.envelope.encode` / `decode`. One broker-neutral metadata
//! contract for every rail (canon 0002): an envelope is its canonical bytes,
//! an incomplete one is refused with its missing fields NAMED (never guessed
//! at), an unknown specversion is refused by name, and unknown additive
//! fields survive a decode (a newer publisher never breaks an older reader).
//!
//! Minting (`make_envelope`) needs a clock and randomness; it is not sp1's.

use crate::canonical::canonical;
use serde_json::Value;
use std::fmt;

pub const SPECVERSION: &str = "orreth.transport/1";
pub const KINDS: [&str; 2] = ["command", "event"];
pub const REQUIRED: [&str; 8] = [
    "specversion",
    "message_id",
    "message_kind",
    "type",
    "universe_id",
    "scope_path",
    "occurred_at",
    "payload",
];

/// The refusals, worded as the reference words them (the fixtures assert the names).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EnvelopeError {
    /// `envelope missing required fields: a, b` — every missing name, in REQUIRED's order.
    MissingFields(Vec<String>),
    /// `unknown specversion '…'`.
    UnknownSpecversion(String),
    /// The value is not a JSON object.
    NotAnObject,
    /// The bytes are not ASCII JSON.
    NotJson(String),
}

impl fmt::Display for EnvelopeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            EnvelopeError::MissingFields(names) => {
                write!(f, "envelope missing required fields: {}", names.join(", "))
            }
            EnvelopeError::UnknownSpecversion(v) => write!(f, "unknown specversion '{v}'"),
            EnvelopeError::NotAnObject => write!(f, "an envelope is a JSON object"),
            EnvelopeError::NotJson(why) => write!(f, "an envelope is ASCII JSON: {why}"),
        }
    }
}

impl std::error::Error for EnvelopeError {}

fn missing(env: &serde_json::Map<String, Value>) -> Result<(), EnvelopeError> {
    let names: Vec<String> = REQUIRED
        .iter()
        .filter(|k| !env.contains_key(**k))
        .map(|k| k.to_string())
        .collect();
    if names.is_empty() {
        Ok(())
    } else {
        Err(EnvelopeError::MissingFields(names))
    }
}

/// Canonical bytes of a complete envelope; an incomplete one is refused with
/// its missing fields named, a foreign specversion by name.
pub fn encode(env: &Value) -> Result<Vec<u8>, EnvelopeError> {
    let map = env.as_object().ok_or(EnvelopeError::NotAnObject)?;
    missing(map)?;
    let spec = &map["specversion"];
    if spec.as_str() != Some(SPECVERSION) {
        return Err(EnvelopeError::UnknownSpecversion(crate::py::python_str(
            spec,
        )));
    }
    Ok(canonical(env))
}

/// Read an envelope back. Unknown additive fields are preserved; missing
/// required ones refuse.
pub fn decode(raw: &[u8]) -> Result<Value, EnvelopeError> {
    if !raw.is_ascii() {
        return Err(EnvelopeError::NotJson("not ascii".into()));
    }
    let env: Value =
        serde_json::from_slice(raw).map_err(|e| EnvelopeError::NotJson(e.to_string()))?;
    let map = env.as_object().ok_or(EnvelopeError::NotAnObject)?;
    missing(map)?;
    Ok(env)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn full() -> Value {
        json!({
            "specversion": SPECVERSION, "message_id": "msg_1", "message_kind": "event",
            "type": "orreth.heartbeat.v1", "universe_id": "u:dev", "scope_path": "u:dev",
            "occurred_at": "2026-09-22T00:00:00.000Z", "payload": {"ref": "rec_1", "hash": "sha256:-"}
        })
    }

    #[test]
    fn refusals_name_what_is_missing_in_required_order() {
        let mut env = full();
        env.as_object_mut().unwrap().remove("occurred_at");
        env.as_object_mut().unwrap().remove("message_id");
        let err = encode(&env).unwrap_err();
        assert_eq!(
            err.to_string(),
            "envelope missing required fields: message_id, occurred_at"
        );
        let mut env = full();
        env["specversion"] = json!("orreth.transport/9");
        assert_eq!(
            encode(&env).unwrap_err().to_string(),
            "unknown specversion 'orreth.transport/9'"
        );
        assert_eq!(encode(&json!([1])).unwrap_err(), EnvelopeError::NotAnObject);
    }

    #[test]
    fn decode_round_trips_and_keeps_the_unknown() {
        let mut env = full();
        env["future_field"] = json!({"a": 1});
        let bytes = encode(&env).unwrap();
        assert_eq!(decode(&bytes).unwrap(), env);
        assert!(decode(b"{\"x\":\"\xc3\xbc\"}").is_err());
        assert!(decode(b"[]").is_err());
    }
}
